server/scheduler.go

```diff
@@ -297,21 +297,22 @@ func scheduleRemovePeer(cluster *clusterInfo, s Selector, filters ...Filter) (*r
 
 // scheduleTransferLeader schedules a region to transfer leader to the peer.
 func scheduleTransferLeader(cluster *clusterInfo, s Selector, filters ...Filter) (*regionInfo, *metapb.Peer) {
-	sourceStores := cluster.getStores()
-
-	source := s.SelectSource(sourceStores, filters...)
-	if source == nil {
+	region := cluster.randomRegion()
+	if region == nil {
 		return nil, nil
 	}
 
-	region := cluster.randLeaderRegion(source.GetId())
-	if region == nil {
+	leaderStore := cluster.getStore(region.Leader.GetStoreId())
```

Comment:
@nolouch Assume the peers are perfectly balanced, only leaders need to transfer. The rate of successfully find a region that need be balanced is mainly determined by node count and how many nodes are not well balanced.
Assume we have 100000 regions and 10 nodes, then each node contains 10000 leaders. When a region is randomly selected, the rate of its leader on a specific node is 10%.
I have checked that it works well on a 10 nodes cluster, it may be inefficient it if there are more nodes. I'll continue do more test :)